### PR TITLE
fix: add `userMatrix` property to generic actor serializer

### PR DESCRIPTION
**Pull request by FelipeCybis**
- Solves https://github.com/Kitware/trame/issues/336

Add `GetUserMatrix` property to `generic_actor_serializer` and `generic_volume_serializer`.

### CODE REVIEW

**Comment by FelipeCybis** [comment]
I don't know if there is any CI I should have done... it works in my machine :) tell me if I can do something more.

**Comment by jourdain** [comment]
then run black or the pre-commit hooks.

**Comment by jourdain** [comment]
I'm guessing it is still working as expected on your end. Also do you know if we do have some use case where the user matrix is not getting extracted? Just double checking

**Comment by FelipeCybis** [comment]
hmm, I can check tomorrow with `vtkVolume` if it still work as expected, but I don't have an extensive knowledge on vtk to be able to think in more use cases. I am happy to test some things out if you have ideas.

Also here we pass only the user matrix... if I understood well the actor will also use `UserTransform` on remote, if it has any... so this is missing.

**Comment by jourdain** [comment]
I will look at the tests to see why they are failing... And try to make sure we are properly skipping the matrix if not used/set.

**Comment by FelipeCybis** [comment]
Will do, but even before, if the matrix was none, the identity was passed.

**Comment by jourdain** [comment]
Actually if you can share a tiny pyvista(or vtk) example making use of it, that would be great.

**Comment by FelipeCybis** [comment]
Here are the codes I am using:

## vtkActor

``` python
import os
import numpy as np

from trame.app import get_server
from trame.ui.vuetify import SinglePageWithDrawerLayout
from trame.widgets import vtk, vuetify

from vtkmodules.vtkCommonMath import vtkMatrix4x4
from vtkmodules.vtkCommonDataModel import vtkDataObject
from vtkmodules.vtkFiltersCore import vtkContourFilter
from vtkmodules.vtkIOXML import vtkXMLUnstructuredGridReader
from vtkmodules.vtkRenderingAnnotation import vtkCubeAxesActor

from vtkmodules.vtkRenderingCore import (
    vtkActor,
    vtkDataSetMapper,
    vtkRenderer,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
)

# Required for interactor initialization
from vtkmodules.vtkInteractionStyle import vtkInteractorStyleSwitch  # noqa

# Required for rendering initialization, not necessary for
# local rendering, but doesn't hurt to include it
import vtkmodules.vtkRenderingOpenGL2  # noqa

CURRENT_DIRECTORY = os.path.abspath(os.path.dirname(__file__))

# -----------------------------------------------------------------------------
# Constants
# -----------------------------------------------------------------------------


class Representation:
    Points = 0
    Wireframe = 1
    Surface = 2
    SurfaceWithEdges = 3


class LookupTable:
    Rainbow = 0
    Inverted_Rainbow = 1
    Greyscale = 2
    Inverted_Greyscale = 3


# -----------------------------------------------------------------------------
# VTK pipeline
# -----------------------------------------------------------------------------

renderer = vtkRenderer()
renderWindow = vtkRenderWindow()
renderWindow.AddRenderer(renderer)

renderWindowInteractor = vtkRenderWindowInteractor()
renderWindowInteractor.SetRenderWindow(renderWindow)
renderWindowInteractor.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

# Read Data
reader = vtkXMLUnstructuredGridReader()
reader.SetFileName(os.path.join(CURRENT_DIRECTORY, "../data/disk_out_ref.vtu"))
reader.Update()

# Extract Array/Field information
dataset_arrays = []
fields = [
    (reader.GetOutput().GetPointData(), vtkDataObject.FIELD_ASSOCIATION_POINTS),
    (reader.GetOutput().GetCellData(), vtkDataObject.FIELD_ASSOCIATION_CELLS),
]
for field in fields:
    field_arrays, association = field
    for i in range(field_arrays.GetNumberOfArrays()):
        array = field_arrays.GetArray(i)
        array_range = array.GetRange()
        dataset_arrays.append(
            {
                "text": array.GetName(),
                "value": i,
                "range": list(array_range),
                "type": association,
            }
        )
default_array = dataset_arrays[0]
default_min, default_max = default_array.get("range")

# Mesh
mesh_mapper = vtkDataSetMapper()
mesh_mapper.SetInputConnection(reader.GetOutputPort())
mesh_actor = vtkActor()
mesh_actor.SetMapper(mesh_mapper)

matrix = vtkMatrix4x4()
# scaling
zoom = np.eye(4)
np.fill_diagonal(zoom[:3,:3], 4)

# translation 
translate = np.eye(4)
translate[:3, 3] = [2, 2, 0]

# rotation
rot = np.eye(4)
rot[1,1] = np.cos(-np.pi/4)
rot[2,2] = np.cos(-np.pi/4)
rot[1,2] = -np.sin(-np.pi/4)
rot[2,1] = np.sin(-np.pi/4)

# shear
shear = np.eye(4)
shear[0, 1] = -1

user_matrix = translate @ rot @ zoom @ shear
m, n = user_matrix.shape
for i in range(m):
    for j in range(n):
        matrix.SetElement(i, j, user_matrix[i, j])
mesh_actor.SetUserMatrix(matrix)

renderer.AddActor(mesh_actor)

# Mesh: Setup default representation to surface
mesh_actor.GetProperty().SetRepresentationToSurface()
mesh_actor.GetProperty().SetPointSize(1)
mesh_actor.GetProperty().EdgeVisibilityOff()

# Mesh: Apply rainbow color map
mesh_lut = mesh_mapper.GetLookupTable()
mesh_lut.SetHueRange(0.666, 0.0)
mesh_lut.SetSaturationRange(1.0, 1.0)
mesh_lut.SetValueRange(1.0, 1.0)
mesh_lut.Build()

# Mesh: Color by default array
mesh_mapper.SelectColorArray(default_array.get("text"))
mesh_mapper.GetLookupTable().SetRange(default_min, default_max)
if default_array.get("type") == vtkDataObject.FIELD_ASSOCIATION_POINTS:
    mesh_mapper.SetScalarModeToUsePointFieldData()
else:
    mesh_mapper.SetScalarModeToUseCellFieldData()
mesh_mapper.SetScalarVisibility(True)
mesh_mapper.SetUseLookupTableScalarRange(True)

# Contour
contour = vtkContourFilter()
contour.SetInputConnection(reader.GetOutputPort())
contour_mapper = vtkDataSetMapper()
contour_mapper.SetInputConnection(contour.GetOutputPort())
contour_actor = vtkActor()
contour_actor.SetMapper(contour_mapper)
renderer.AddActor(contour_actor)

# Contour: ContourBy default array
contour_value = 0.5 * (default_max + default_min)
contour.SetInputArrayToProcess(
    0, 0, 0, default_array.get("type"), default_array.get("text")
)
contour.SetValue(0, contour_value)

# Contour: Setup default representation to surface
contour_actor.GetProperty().SetRepresentationToSurface()
contour_actor.GetProperty().SetPointSize(1)
contour_actor.GetProperty().EdgeVisibilityOff()

# Contour: Apply rainbow color map
contour_lut = contour_mapper.GetLookupTable()
contour_lut.SetHueRange(0.666, 0.0)
contour_lut.SetSaturationRange(1.0, 1.0)
contour_lut.SetValueRange(1.0, 1.0)
contour_lut.Build()

# Contour: Color by default array
contour_mapper.SelectColorArray(default_array.get("text"))
contour_mapper.GetLookupTable().SetRange(default_min, default_max)
if default_array.get("type") == vtkDataObject.FIELD_ASSOCIATION_POINTS:
    contour_mapper.SetScalarModeToUsePointFieldData()
else:
    contour_mapper.SetScalarModeToUseCellFieldData()
contour_mapper.SetScalarVisibility(True)
contour_mapper.SetUseLookupTableScalarRange(True)

# Cube Axes
cube_axes = vtkCubeAxesActor()
renderer.AddActor(cube_axes)

# Cube Axes: Boundaries, camera, and styling
cube_axes.SetBounds(mesh_actor.GetBounds())
cube_axes.SetCamera(renderer.GetActiveCamera())
cube_axes.SetXLabelFormat("%6.1f")
cube_axes.SetYLabelFormat("%6.1f")
cube_axes.SetZLabelFormat("%6.1f")
cube_axes.SetFlyModeToOuterEdges()

renderer.ResetCamera()

# -----------------------------------------------------------------------------
# Trame setup
# -----------------------------------------------------------------------------

server = get_server()
server.client_type = "vue2"
state, ctrl = server.state, server.controller

state.setdefault("active_ui", None)


# -----------------------------------------------------------------------------
# GUI
# -----------------------------------------------------------------------------

with SinglePageWithDrawerLayout(server) as layout:
    layout.title.set_text("Viewer")

    with layout.content:
        with vuetify.VContainer(
            fluid=True,
            classes="pa-0 fill-height",
        ):
            with vuetify.VContainer(
                fluid=True, classes="pa-0 fill-height", style="width: 50%;"
            ):
                local = vtk.VtkLocalView(renderWindow, on_ready="local_rendering_ready++")
            with vuetify.VContainer(
                fluid=True, classes="pa-0 fill-height", style="width: 50%;"
            ):
                remote = vtk.VtkRemoteView(renderWindow)

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------

if __name__ == "__main__":
    server.start()
```

## vtkVolume

``` python
import vtk
from vtkmodules.vtkCommonMath import vtkMatrix4x4

import numpy as np
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import vuetify, html
from trame.widgets.vtk import VtkLocalView, VtkRemoteView

# Just for using this script in testing
# from trame_client.utils.testing import enable_testing

# server = enable_testing(get_server(), "local_rendering_ready")
server = get_server()
server.client_type = "vue2"
state, ctrl = server.state, server.controller

state.trame__title = "VTK Volume Rendering"
state.local_rendering_ready = 0

# MAPPER_TYPE = "FixedPoint"
MAPPER_TYPE = "Smart"
# MAPPER_TYPE = "GPU"
# MAPPER_TYPE = "RayCast"
MAPPERS = {
    "FixedPoint": vtk.vtkFixedPointVolumeRayCastMapper(),
    "Smart": vtk.vtkSmartVolumeMapper(),
    "GPU": vtk.vtkOpenGLGPUVolumeRayCastMapper(),
    "RayCast": vtk.vtkGPUVolumeRayCastMapper(),
}

# -----------------------------------------------------------------------------
ren = vtk.vtkRenderer()
renWin = vtk.vtkRenderWindow()
renWin.AddRenderer(ren)
iren = vtk.vtkRenderWindowInteractor()
iren.SetRenderWindow(renWin)
iren.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

source = vtk.vtkRTAnalyticSource()
source.Update()
mapper = MAPPERS[MAPPER_TYPE]
mapper.SetInputConnection(source.GetOutputPort())
actor = vtk.vtkVolume()
actor.SetMapper(mapper)
actor.GetProperty().SetScalarOpacityUnitDistance(10)

matrix = vtkMatrix4x4()
# scaling
zoom = np.eye(4)
np.fill_diagonal(zoom[:3,:3], 4)

# translation 
translate = np.eye(4)
translate[:3, 3] = [2, 2, 0]

# rotation
rot = np.eye(4)
rot[1,1] = np.cos(-np.pi/4)
rot[2,2] = np.cos(-np.pi/4)
rot[1,2] = -np.sin(-np.pi/4)
rot[2,1] = np.sin(-np.pi/4)

# shear
shear = np.eye(4)
shear[0, 1] = -1

user_matrix = translate @ rot @ zoom
m, n = user_matrix.shape
for i in range(m):
    for j in range(n):
        matrix.SetElement(i, j, user_matrix[i, j])

actor.SetUserMatrix(matrix)


ren.AddActor(actor)

colorTransferFunction = vtk.vtkColorTransferFunction()
colorTransferFunction.AddRGBPoint(0.0, 0.0, 0.0, 0.0)
colorTransferFunction.AddRGBPoint(64.0, 1.0, 0.0, 0.0)
colorTransferFunction.AddRGBPoint(128.0, 0.0, 0.0, 1.0)
colorTransferFunction.AddRGBPoint(192.0, 0.0, 1.0, 0.0)
colorTransferFunction.AddRGBPoint(255.0, 0.0, 0.2, 0.0)

opacityTransferFunction = vtk.vtkPiecewiseFunction()
opacityTransferFunction.AddPoint(20, 0.0)
opacityTransferFunction.AddPoint(255, 0.2)

volumeProperty = vtk.vtkVolumeProperty()
volumeProperty.SetColor(colorTransferFunction)
volumeProperty.SetScalarOpacity(opacityTransferFunction)
volumeProperty.ShadeOn()
volumeProperty.SetInterpolationTypeToLinear()

actor.SetProperty(volumeProperty)

cube = vtk.vtkCubeAxesActor()
cube.SetCamera(ren.GetActiveCamera())
cube.SetBounds(actor.GetBounds())

ren.AddActor(cube)

iren.Initialize()
ren.ResetCamera()
ren.SetBackground(0.7, 0.7, 0.7)
renWin.Render()


# -----------------------------------------------------------------------------
# GUI
# -----------------------------------------------------------------------------

with SinglePageLayout(server) as layout:
    layout.icon.click = ctrl.view_reset_camera
    layout.title.set_text(state.trame__title)

    with layout.toolbar:
        vuetify.VSpacer()
        html.Div("{{ local_rendering_ready }}", classes="readyCount")

    with layout.content:
        with vuetify.VContainer(
            fluid=True,
            classes="pa-0 fill-height",
        ):
            with vuetify.VContainer(
                fluid=True, classes="pa-0 fill-height", style="width: 50%;"
            ):
                local = VtkLocalView(renWin, on_ready="local_rendering_ready++")
            with vuetify.VContainer(
                fluid=True, classes="pa-0 fill-height", style="width: 50%;"
            ):
                remote = VtkRemoteView(renWin)

    # hide footer
    layout.footer.hide()

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------

if __name__ == "__main__":
    server.start()

```

## Pyvista volume
```python
"""Validate volume rendering with VTK.js."""
import pyvista as pv
import numpy as np
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import vuetify, html
from trame.widgets.vtk import VtkLocalView, VtkRemoteView

server = get_server()
server.client_type = "vue2"
state, ctrl = server.state, server.controller

state.trame__title = "Volume Validation"
state.local_rendering_ready = 0

# -----------------------------------------------------------------------------

image = pv.Wavelet()

# scaling
zoom = np.eye(4)
np.fill_diagonal(zoom[:3,:3], 4)

# translation 
translate = np.eye(4)
translate[:3, 3] = [20, 20, 0]

# rotation
rot = np.eye(4)
rot[1,1] = np.cos(-np.pi/4)
rot[2,2] = np.cos(-np.pi/4)
rot[1,2] = np.sin(-np.pi/4)
rot[2,1] = -np.sin(-np.pi/4)

# shear
shear = np.eye(4)
shear[0, 1] = -1

plotter = pv.Plotter(off_screen=True)
actor = plotter.add_volume(image, mapper="smart", user_matrix=translate @ zoom @ shear)
plotter.show_grid()
plotter.reset_camera()


def update_local_rendering():
    ctrl.view_update()
    ctrl.view_reset_camera()


# -----------------------------------------------------------------------------
# GUI
# -----------------------------------------------------------------------------

with SinglePageLayout(server) as layout:
    layout.icon.click = ctrl.view_reset_camera
    layout.title.set_text(state.trame__title)

    with layout.toolbar:
        vuetify.VSpacer()
        html.Div("{{ local_rendering_ready }}", classes="readyCount")
        vuetify.VBtn("Update", click=update_local_rendering)

    with layout.content:
        with vuetify.VContainer(
            fluid=True,
            classes="pa-0 fill-height",
        ):
            with vuetify.VCol(classes="fill-height"):
                view = VtkLocalView(
                    plotter.ren_win,
                    ref="local",
                    on_ready="local_rendering_ready++",
                )
                ctrl.view_update = view.update
                ctrl.view_reset_camera = view.reset_camera
            with vuetify.VCol(classes="fill-height"):
                VtkRemoteView(
                    plotter.ren_win,
                    ref="remote",
                )

    # hide footer
    layout.footer.hide()

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------

if __name__ == "__main__":
    server.start()

```

**Comment by FelipeCybis** [comment]
For the mesh, it works fine (even though I don't know what is happening with the colormap): 
image

But there is effectively some weird stuff going on on volume rendering. It works as intended with zoom and translations, but shear and rotations render weird volumes:

image

Since the Prop3D in vtk-js is parent for both vtkActor and vtkVolume, I don't really know what is going on. I would like to test directly on a pure vtk-js application, but I would need a few days in the "Intro to vtk-js" part.

**Comment by FelipeCybis** [comment]
It turns out I do care about this :P

- https://github.com/Kitware/vtk-js/issues/2922

File: trame_vtk/modules/vtk/serializers/actors.py [hunk 1]
```diff
@@ -60,6 +61,16 @@ def generic_actor_serializer(parent, actor, actor_id, context, depth):
                 dependencies.append(texture_instance)
                 calls.append(["addTexture", [wrap_id(texture_id)]])
 
+    # Apply transform
+    if actor.GetUserMatrix():
+        user_matrix = [0] * 16
+        matrix = actor.GetUserMatrix()
+        for i in range(4):
+            for j in range(4):
+                idx = i + 4 * j
```
Comment (jourdain) [inline trame_vtk/modules/vtk/serializers/actors.py:70]: Not sure but we may need to transpose the matrix due to different storage in vtk.js vs VTK.

Comment (FelipeCybis) [reply trame_vtk/modules/vtk/serializers/actors.py:70]: normally the for loop is doing it, vtk.js uses column major raveled matrices if I understood it well, so `matrix.GetElement(i,j)` is element `i + 4*j`. The mesh transformation show the serialization is good.

**Comment by FelipeCybis** [comment]
Hey @jourdain, it turns out the vtk.js problem was solved in v29.1.1 https://github.com/Kitware/vtk-js/pull/2954, but I cannot figure it out if this update is automatic when trame uses the local view. If this is really settled in vtk.js side, it would be cool to figure it out how to merge this.

**Comment by jourdain** [comment]
yes, we just need to trigger a new release of vue-vtk-js with the latest vtk.js. 
Then pick that version in trame-vtk.

**Comment by jourdain** [comment]
(I just merged the vtk.js update)

**Comment by FelipeCybis** [comment]
I did rebase the branch, but still having the same issues with the Volume rendering (local rendering on the left, remote on the right)

image

Mesh actors are still good. I am not sure I am serializing correctly, or in the correct function. I guess so, but the bump to vtk-js 29.1.1 do not seem to have made any difference here, while I can explicitly use the userMatrix on a pure vtk-js application, using:

```html
<!DOCTYPE html>
<html>
<body>
<!-- <script type="text/javascript" src="https://unpkg.com/vtk.js@28.8"></script> -->
<script type="text/javascript" src="https://unpkg.com/vtk.js@29.1.1"></script>
<script type="text/javascript">
  // --------------------------------------------------------------------------
  // Example code
  // --------------------------------------------------------------------------
  var fullScreenRenderer = vtk.Rendering.Misc.vtkFullScreenRenderWindow.newInstance();
  var actor              = vtk.Rendering.Core.vtkVolume.newInstance();
  var mapper             = vtk.Rendering.Core.vtkVolumeMapper.newInstance();
  var vol               = vtk.Filters.Sources.vtkRTAnalyticSource.newInstance();
  var actor_2              = vtk.Rendering.Core.vtkVolume.newInstance();
  var mapper_2             = vtk.Rendering.Core.vtkVolumeMapper.newInstance();
  var vol_2               = vtk.Filters.Sources.vtkRTAnalyticSource.newInstance();

  actor.setMapper(mapper);
  mapper.setInputConnection(vol.getOutputPort());

  actor_2.setMapper(mapper_2);
  mapper_2.setInputConnection(vol_2.getOutputPort());

  var renderer = fullScreenRenderer.getRenderer();
  actor.setUserMatrix([1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1])
  actor_2.setUserMatrix([3,0,0,0,3,2.12132,-2.12132,0,0,2.12132,2.12132,0,100,0,0,1])

  renderer.addActor(actor);
  renderer.addActor(actor_2);
  renderer.resetCamera();

  var renderWindow = fullScreenRenderer.getRenderWindow();
  renderWindow.render();
</script>
</body>
</html>
```
image

I am not sure what I am missing here...

**Comment by jourdain** [comment]
Did you try to clear your browser cache? I just looked at the code and I don't see anything wrong especially if that was indeed fixed in vtk.js.

Thanks for your contribution !

**Comment by FelipeCybis** [comment]
Yep, something as dumb as this. Thanks for trame :)

**Comment by jourdain** [comment]
I'm glad you like trame and you see value in it.

**Pull request merged**